keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,549 @@
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+
+
+## Motivation
+
+High Performance Computing / Machine Learning users usually want some control over how jobs are being created.  A Startup policy would allow a user to specify which status a replicated job should have before starting other replicated jobs in a jobset.
+
+### Goals
+
+- Add a startup policy for JobSet
+
+### Non-Goals
+
+- Complicated workflow specifications
+  - This means that we will support simple dependecies rather than processing a full dag.
+  - Projects like argo-workflows are a better use case for this.
+- Starting Jobs based on Ready will be at the ReplicatedJob level not at the Job level.
+  - ReplicatedJobs will be classified as ready if all jobs are ready.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, I have one ReplicatedJob that functions as the driver and a series of worker ReplicatedJobs.  
+I want the driver to be ready before I start the worker replicated jobs.  
+This is useful for HPC/AI/ML workloads as some frameworks have a driver pod that must be ready before the workers are started.  
+
+The API below would fit my purpose.
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: driver-ready-worker-start
+spec:
+  startupPolicy:
+  - name: driver
+    status: Ready
+  replicatedJobs:
+  - name: driver
+    replicas: 1
+    template:
+      spec:
+        # Set backoff limit to 0 so job will immediately fail if any pod fails.
+        backoffLimit: 0 
+        completions: 1
+        parallelism: 1
+        template:
+          spec:
+            containers:
+            - name: driver
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10000
+  - name: workers
+    replicas: 1
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+```
+
+#### Story 2
+
+As a user, I have one ReplicatedJob that I want to finish before going on to the next worker.
```

Comment:
I think we need to be careful about turning jobset into a workflow manager. Workflow managers are much more than simple sequencing, and I am concerned that we will be offering a half-baked solution with this approach.

Reply:
Agreed.  I already had some comments in this KEP/issue about this.  I was thinking that this could be fine as we only support sequential and no dag syntax.  

Similar to init containers (start in order).  It gets complicated because if you specify 1 or 2 jobs, what do you do with the others?

Not sure hence why I thought a KEP would be useful at first.